File: xkcd/src/main/java/xyz/jienan/xkcd/whatif/presenter/WhatIfMainPresenter.java
package xyz.jienan.xkcd.whatif.presenter;

import java.util.Collections;
import java.util.List;
import java.util.Random;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import timber.log.Timber;
import xyz.jienan.xkcd.model.WhatIfArticle;
import xyz.jienan.xkcd.model.WhatIfModel;
import xyz.jienan.xkcd.model.persist.SharedPrefManager;
import xyz.jienan.xkcd.whatif.contract.WhatIfMainContract;

public class WhatIfMainPresenter implements WhatIfMainContract.Presenter {

    private final SharedPrefManager sharedPrefManager = SharedPrefManager.INSTANCE;

    private WhatIfMainContract.View view;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private Disposable fabShowDisposable = Disposables.empty();

    private Disposable searchDisposable = Disposables.empty();

    private WhatIfMainPresenter() {
        // no default public constructor
    }

    public WhatIfMainPresenter(WhatIfMainContract.View view) {
        this.view = view;
    }

    @Override
    public void favorited(long index, boolean isFav) {
        if (index < 1) {
            return;
        }
        Disposable d = WhatIfModel.INSTANCE.fav(index, isFav).subscribe(xkcdPic -> {
                },
                e -> Timber.e(e, "error on get one pic: %d", index));
        compositeDisposable.add(d);
        view.toggleFab(isFav);
    }

    @Override
    public void liked(long currentIndex) {
        if (currentIndex < 1) {
            return;
        }
        compositeDisposable.add(
                WhatIfModel.INSTANCE
                        .thumbsUp(currentIndex)
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(view::showThumbUpCount,
                                e -> Timber.e(e, "Thumbs up failed")));
    }

    @Override
    public void getInfoAndShowFab(int index) {
        if (fabShowDisposable != null && !fabShowDisposable.isDisposed()) {
            fabShowDisposable.dispose();
        }
        WhatIfArticle article = WhatIfModel.INSTANCE.loadArticleFromDB(index);
        if (article == null) {
            fabShowDisposable = WhatIfModel.INSTANCE.observe()
                    .filter(article1 -> article1.num == index)
                    .subscribe(view::showFab,
                            e -> Timber.e(e, "what if pipeline observing error"));
            compositeDisposable.add(fabShowDisposable);
        } else {
            view.showFab(article);
        }
    }

    @Override
    public void setLastViewed(int lastViewed) {
        sharedPrefManager.setLastViewedWhatIf(lastViewed);
    }

    @Override
    public int getLatest() {
        return (int) sharedPrefManager.getLatestWhatIf();
    }

    @Override
    public void setLatest(int latestIndex) {
        sharedPrefManager.setLastViewedWhatIf(latestIndex);
    }

    @Override
    public int getLastViewed(int latestIndex) {
        return (int) sharedPrefManager.getLastViewedWhatIf(latestIndex);
    }

    @Override
    public void loadLatest() {
        Disposable d = WhatIfModel.INSTANCE.loadLatest()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(whatIfArticle -> {
                    long latestIndex = whatIfArticle.num;
                    sharedPrefManager.setLatestWhatIf(latestIndex);
                    view.latestWhatIfLoaded(whatIfArticle);
                }, e -> Timber.e(e, "load what if article error"));
        compositeDisposable.add(d);
    }

    @Override
    public void searchContent(String query) {

        if (!searchDisposable.isDisposed()) {
            searchDisposable.dispose();
        }

        searchDisposable = WhatIfModel.INSTANCE.searchWhatIf(query, sharedPrefManager.getWhatIfSearchPref())
                .map(list -> {
                    if (isNumQuery(query)) {
                        long num = Long.parseLong(query);
                        WhatIfArticle matchNumArticle = null;
                        for (WhatIfArticle article : list) {
                            if (article.num == num) {
                                matchNumArticle = article;
                                break;
                            }
                        }
                        if (matchNumArticle != null) {
                            list.remove(matchNumArticle);
                            list.add(0, matchNumArticle);
                        }
                    }
                    return list;
                })
                .subscribe(view::renderWhatIfSearch,
                        e -> {
                    Timber.e(e, "search what if error");
                    if (isNumQuery(query)) {
                        long num = Long.parseLong(query);
                        WhatIfArticle article = WhatIfModel.INSTANCE.loadArticleFromDB(num);
                        if (article != null) {
                            view.renderWhatIfSearch(Collections.singletonList(article));
                        }
                    }
                });
    }

    @Override
    public void onDestroy() {
        compositeDisposable.dispose();
        fabShowDisposable.dispose();
        searchDisposable.dispose();
    }

    @Override
    public long getRandomUntouchedIndex() {
        final List<WhatIfArticle> list = WhatIfModel.INSTANCE.getUntouchedList();
        if (list.isEmpty()) {
            return 0;
        } else {
            return list.get(new Random().nextInt(list.size())).num;
        }
    }

    private boolean isNumQuery(String query) {
        try {
            long num = Long.parseLong(query);
            return num > 0 && num <= sharedPrefManager.getLatestWhatIf();
        } catch (NumberFormatException e) {
            return false;
        }
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/whatif/presenter/WhatIfMainPresenter.kt
package xyz.jienan.xkcd.whatif.presenter

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposables
import timber.log.Timber
import xyz.jienan.xkcd.Const.WHAT_IF_BOOKMARK
import xyz.jienan.xkcd.model.WhatIfArticle
import xyz.jienan.xkcd.model.WhatIfModel
import xyz.jienan.xkcd.model.persist.SharedPrefManager
import xyz.jienan.xkcd.whatif.contract.WhatIfMainContract
import java.util.concurrent.TimeUnit
import kotlin.random.Random

class WhatIfMainPresenter constructor(private val view: WhatIfMainContract.View) : WhatIfMainContract.Presenter {

    private val compositeDisposable = CompositeDisposable()

    private var fabShowDisposable = Disposables.empty()

    private var searchDisposable = Disposables.empty()

    override fun favorited(currentIndex: Long, isFav: Boolean) {
        if (currentIndex < 1) {
            return
        }
        WhatIfModel.fav(currentIndex, isFav).subscribe({},
                { e -> Timber.e(e, "error on get one pic: %d", currentIndex) })
                .also { compositeDisposable.add(it) }
        view.toggleFab(isFav)
    }

    override fun liked(currentIndex: Long) {
        if (currentIndex < 1) {
            return
        }

        WhatIfModel.thumbsUp(currentIndex)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ view.showThumbUpCount(it) },
                        { e -> Timber.e(e, "Thumbs up failed") })
                .also { compositeDisposable.add(it) }
    }

    override fun getInfoAndShowFab(index: Int) {
        if (!fabShowDisposable.isDisposed) {
            fabShowDisposable.dispose()
        }
        val article = WhatIfModel.loadArticleFromDB(index.toLong())
        if (article == null) {
            fabShowDisposable = WhatIfModel.observe()
                    .filter { (num) -> num == index.toLong() }
                    .subscribe({ view.showFab(it) },
                            { e -> Timber.e(e, "what if pipeline observing error") })
            compositeDisposable.add(fabShowDisposable)
        } else {
            view.showFab(article)
        }
    }

    override fun setLastViewed(lastViewed: Int) {
        SharedPrefManager.setLastViewedWhatIf(lastViewed.toLong())
    }

    override var latest: Int
        get() = SharedPrefManager.latestWhatIf.toInt()
        set(value) {
            SharedPrefManager.setLastViewedWhatIf(value.toLong())
        }

    override fun getLastViewed(latestIndex: Int) =
            SharedPrefManager.getLastViewedWhatIf(latestIndex.toLong()).toInt()

    override fun loadLatest() {
        WhatIfModel.loadLatest()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ whatIfArticle ->
                    val latestIndex = whatIfArticle.num
                    SharedPrefManager.latestWhatIf = latestIndex
                    view.latestWhatIfLoaded(whatIfArticle)
                }, { e -> Timber.e(e, "load what if article error") })
                .also { compositeDisposable.add(it) }
    }

    override fun searchContent(query: String) {

        if (!searchDisposable.isDisposed) {
            searchDisposable.dispose()
        }

        // Can't subscribe on io thread and observe on main thread, otherwise cursor behaves weirdly
        searchDisposable = WhatIfModel.searchWhatIf(query, SharedPrefManager.whatIfSearchPref)
                .map { list ->
                    if (isNumQuery(query)) {
                        val num = query.toLong()
                        moveNumberQueryToFirstPlace(list, num)
                    }
                    list
                }
                .subscribe({ view.renderWhatIfSearch(it) },
                        { e ->
                            Timber.e(e, "search what if error")
                            fallbackIfNumberQuery(query)
                        })
    }

    override fun getBookmark() = SharedPrefManager.getBookmark(WHAT_IF_BOOKMARK)

    override fun setBookmark(index: Long) : Boolean {
        return if (index > 0) {
            SharedPrefManager.setBookmark(WHAT_IF_BOOKMARK, index)
            true
        } else {
            false
        }
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
        fabShowDisposable.dispose()
        searchDisposable.dispose()
    }

    override val randomUntouchedIndex: Long
        get() {
            val list = WhatIfModel.untouchedList
            return if (list.isEmpty()) {
                0
            } else {
                list[Random.nextInt(list.size)].num
            }
        }

    private fun moveNumberQueryToFirstPlace(list: MutableList<WhatIfArticle>, num: Long) {
        val matchNumArticle = list.firstOrNull { it.num == num }

        if (matchNumArticle != null) {
            list.remove(matchNumArticle)
            list.add(0, matchNumArticle)
        }
    }

    private fun fallbackIfNumberQuery(query: String) {
        if (isNumQuery(query)) {
            val num = query.toLong()
            val article = WhatIfModel.loadArticleFromDB(num)
            if (article != null) {
                view.renderWhatIfSearch(listOf(article))
            }
        }
    }

    private fun isNumQuery(query: String) =
            try {
                val num = query.toLong()
                num > 0 && num <= SharedPrefManager.latestWhatIf
            } catch (e: NumberFormatException) {
                false
            }
}
